rfd/0143-external-k8s-joining.md

```diff
@@ -0,0 +1,280 @@
+---
+authors: Noah Stride (noah.stride@goteleport.com)
+state: draft
+---
+
+# RFD 00143 - External Kubernetes delegated joining (jwks)
+
+## Required approvers
+
+* Engineering: @zmb3 && (@hugoShaka || @tigrato)
+* Product: @klizhentas || @xinding33
+* Security: @reedloden
+
+## What
+
+This RFD proposes a method for allowing entities (such as a Machine ID instance
+or an Agent) to join Teleport Clusters via federation between the Kubernetes 
+Cluster they reside in and a Teleport Cluster, where that Teleport Cluster does 
+not reside in the same Kubernetes Cluster.
+
+This is distinct from 
+[RFD 0094 - Kubernetes node joining](./0094-kubernetes-node-joining.md) which
+only supports the joining entity existing within the same Kubernetes Cluster
+as the Teleport Cluster.
+
+## Why
+
+It is not unusual for users to wish to deploy Machine ID bots and Teleport
+Agents in Kubernetes Clusters which is not the same cluster in which their
+Auth Service resides. For example, they may operate many Kubernetes Clusters or
+they may use Teleport Cloud (in which we host their Teleport Cluster).
+
+Currently, these users must use a different join method, and, for users
+deploying into environments where there is no platform-specific delegated
+join method, they must fallback to using the `token` join method which is a
+sensitive, long-lived, secret. This also requires some element of state, and
+is inconvenient in more ephemeral use-cases.
+
+To provide a more concrete use-case, many users currently deploy Teleport
+Plugins (such as the Slack Access Request plugin). Adding support for External
+Kubernetes Joining will provide a golden pathway for this deployment using
+Machine ID that avoids long-lived shared secrets and provides a seamless 
+experience.
+
+## Details
+
+### Configuration
+
+This join method will be added to the existing `ProvisionTokenV2` specification.
+
+In order to clearly distinguish this new join method from the existing
+`kubernetes` join method, this method will be named `kubernetes-jwks`.
```

Comment:
non blocking: we already discussed naming, but I feel like naming the join method after what the user wants to do instead of what it is make things simpler. I'm reposting this to see if this is a shared feeling.

In my opinion, `kubernetes`(or `kubernetes-internal`)/`kubernetes-remote` has a better affordance than `kubernetes`/`kubernetes-jwks`

Reply:
Yeah - I'm open to thoughts on naming here - I proceeded with `kubernetes-jwks` purely just so I could start writing this RFD. I'm not sure I want to delve into renaming an existing JoinMethod though.

I do agree that naming based on intent, rather than the technical implementation, is going to be easier for users to understand.

I'm not entirely opposed to `kubernetes-remote`, although, as expressed previously, it does make it sound like this join method could not be used for a cluster the Auth Server is running in. But perhaps I'm being overly concerned about this. I'm happy to proceed with `kubernetes-remote` unless @tigrato or the Product folks have a better suggestion.

Reply:
I would say that JWKS is a pretty familiar term to anyone experienced with OIDC providers and JWT, so `kubernetes-jwks` might well be a little more descriptive as to how the functionality actually works under the hood. I am also fine with `kubernetes-remote` though.

Reply:
eaadb1613d